### PR TITLE
add mixture of softmaxes language model

### PR DESCRIPTION
**Pull request by congchan**
## Description ##
Add mixture of softmaxes language model by Zhilin Yang, Zihang Dai, Ruslan Salakhutdinov, William W. Cohen. https://github.com/zihangdai/mos

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-194/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-194/1/index.html

**Comment by congchan** [comment]
Hi, 

The authors claimed it to be better than AWD model. 
Personally I have several experiments, reaching 53 ppl on Test of Penn Treebank using sgd without fine tuning (better than AWD's w/o finetune result - 58 ppl). The model seems promising to me. Will test on WikiText-2 later.

For pre-trained model, what is the specific requirements?

**Comment by mli** [comment]
Job PR-194/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-194/5/index.html

**Comment by mli** [comment]
Job PR-194/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-194/10/index.html

**Comment by mli** [comment]
Job PR-194/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-194/12/index.html

**Comment by cgraywang** [comment]
A quick question, are you able to reproduce the numbers reported in the https://arxiv.org/pdf/1711.03953.pdf either on wikitext-2 or pentreebank? Or even with standard RNN not AWD?

**Comment by congchan** [comment]
@cgraywang I have not yet successfully reproduced resutls

File: gluonnlp/model/train/language_model.py [hunk 1]
```diff
@@ -25,6 +25,157 @@
...
+        self._drop_e = drop_e
+        self._drop_l = drop_l
+        self._num_experts = num_experts
+        self._weight_drop = weight_drop
+        self._tie_weights = tie_weights
+
+        with self.name_scope():
+            self.embedding = self._get_embedding()
+            self.encoder = self._get_encoder()
+            self.prior = nn.Dense(num_experts, use_bias=False, flatten=False)
```
Comment (cgraywang) [inline gluonnlp/model/train/language_model.py]: It would be clearer if specifying the input size: `self.prior = nn.Dense(num_experts, in_units=hidden_size_last , use_bias=False, flatten=False)`

File: gluonnlp/model/train/language_model.py [hunk 2]
```diff
@@ -25,6 +25,157 @@
...
+            out_states.append(state)
+            if self._drop_h and i != len(self.encoder)-1:
+                encoded = nd.Dropout(encoded, p=self._drop_h, axes=(0,))
+                encoded_dropped.append(encoded)
+        if self._dropout:
+            encoded = nd.Dropout(encoded, p=self._dropout, axes=(0,))
+        encoded_dropped.append(encoded)
+        latent = nd.Dropout(self.latent(encoded), p=self._drop_l, axes=(0,))
+        logit = self.decoder(latent.reshape(-1, self._embed_size))
+        prior = nd.softmax(self.prior(encoded).reshape(-1, self._num_experts))
```
Comment (cgraywang) [inline gluonnlp/model/train/language_model.py]: Please separate the `prior_logit` out

File: gluonnlp/model/train/language_model.py [hunk 3]
```diff
@@ -25,6 +25,157 @@
...
+        self._drop_l = drop_l
+        self._num_experts = num_experts
+        self._weight_drop = weight_drop
+        self._tie_weights = tie_weights
+
+        with self.name_scope():
+            self.embedding = self._get_embedding()
+            self.encoder = self._get_encoder()
+            self.prior = nn.Dense(num_experts, use_bias=False, flatten=False)
+            self.latent = nn.Dense(num_experts * embed_size, 'tanh', flatten=False)
```
Comment (cgraywang) [inline gluonnlp/model/train/language_model.py]: It would be clearer if specifying the input size: `self.latent = nn.Dense(num_experts * embed_size, in_units=hidden_size_last, 'tanh', flatten=False)`

File: gluonnlp/model/train/language_model.py [hunk 4]
```diff
@@ -25,6 +25,157 @@
 from gluonnlp.model.utils import _get_rnn_layer
 from gluonnlp.model.utils import apply_weight_drop
 
+class MOSRNN(Block):
```
Comment (cgraywang) [inline gluonnlp/model/train/language_model.py]: Please add test cases to test the forward of the network, especially on the last mixture softmax layer.

**Comment by cgraywang** [comment]
@congchan what's the current number for AWD-LSTM+MoS?

**Comment by congchan** [comment]
@cgraywang   Thanks for the reviews. 

For reproducing results, using sgd with the same hyperparameters suggested by the author https://github.com/zihangdai/mos, my result for the test ppl on penn data is just 61. 

Did not run a full experiments on wikitext-2 as the model is really huge.

**Comment by cgraywang** [comment]
Yes, it looks there is still a consirable gap. Anyway, are you able to use the their code to reproduce their results? Probably we can validate the implementation by comparing with the reference code.

**Comment by szha** [comment]
The clip_global_norm function should only be called on one copy of the all-reduced gradients.

File: gluonnlp/model/train/language_model.py [hunk 5]
```diff
@@ -25,6 +25,157 @@
...
+    drop_i : float
+        Dropout rate to on the output of embedding.
+    drop_e : float
+        Dropout rate to use on the embedding layer.
+    drop_l : float
+        Dropout rate to use on the latent layer.
+    num_experts : int
+        Number of softmax.
+    """
+    def __init__(self, mode, vocab_size, embed_size=280, hidden_size=960, hidden_size_last=620,
```
[the remark this answers is not in the document]
Comment (congchan) [reply gluonnlp/model/train/language_model.py]: The MoS's repo are using different parameters than the AWD parameters. Shall we still use the AWD parameters instead?

**Comment by congchan** [comment]
Just trained a small MoS model with 75.71/72.00 val/test ppl on wikitext-2.
Hyperparameters: emb_size 200, hiden size 600, last hidden size 400, 7 mixture of softmaxes, ...

**Comment by szha** [comment]
@congchan I'd like to take this patch in a feature branch first so that you or others can continue working on this. Thanks for the contribution!

**Comment by cgraywang** [comment]
Thanks @congchan , let's see how to catch up with the PPL on Wikitext-2 reported in the paper later before merging into the master.
